I2C_eeprom_cyclic_store.h

```diff
@@ -0,0 +1,239 @@
+            _isInitialized = true;
+        }
+
+        /**
+         * Read data from the eeprom into an instance. The data is read from
+	 * the current slot of the eeprom.
+         * 
+         * @param buffer the reference to an object to read data for
+         * @return true if data was read successfully, false otherwise
+         */
+        bool read(T& buffer) const { return read(&buffer); }
+
+        /**
+         * Read data from the eeprom into a memory instance. The data is read
+	 * from the current slot of the eeprom.
+         * @param buffer the instance to read data into
+         * @return true if data was read successfully, false otherwise
+         */
+        bool read(T* buffer) const
+        {
+            if(!_isInitialized)
+                return false;
+
+            if(_isEmpty)
+                return false;
+
+            Serial.println("Reading "+String(sizeof(T))+" bytes from eeprom");
+
+            return _eeprom->readBlock((_currentSlot*_bufferPages*_pageSize)+sizeof(_currentVersion), (uint8_t *) buffer, sizeof(T)) == sizeof(T);
+        }
+
+        /**
+         * Write data of object to the next slot in the eeprom. This updates
+	 * the current slot of this instance.
+         * 
+         * @param buffer the reference to the object to write data for
+         * @return true if data was written successfully, false otherwise
+         */
+        bool write(T& buffer) { return write(&buffer); }
+
+        /**
+         * Write data of object to the next slot in the eeprom. This updates
+	 * the current slot of the instance.
+         *
+         * @param buffer the pointer to the instance to write data for
+         * @return true if data was written successfully, false otherwise
+         */
+        bool write(T* buffer)
+        {
+            if(!_isInitialized)
+            {
+                return false;
+            }
+
+            if(_isEmpty)
+            {
+                _currentSlot = 0;
+                _currentVersion = 0;
+            }
+            else
+            {
+                _currentSlot++;
+                _currentVersion++;
+
+                // Wrap around to start if going past end of alotted region
+                uint16_t maxSlots = _totalPages/_bufferPages;
+                if(_currentSlot > maxSlots)
+                    _currentSlot = 0;
+            }
+
+            auto buffer_length = sizeof(_currentVersion)+sizeof(T);
+            uint8_t tmp[buffer_length];
+
+            memcpy(tmp, &_currentVersion, sizeof(_currentVersion));
+            memcpy(tmp+sizeof(_currentVersion), buffer, sizeof(T));
+
+            Serial.println("Writing "+String(buffer_length)+" bytes to eeprom");
+
+            auto success = _eeprom->writeBlock(_currentSlot*_bufferPages*_pageSize, tmp, buffer_length) == 0;
+
+            if(success)
+                _isEmpty = false;
+
+            return success;
+        }   
+
+    private:
+        uint16_t _pageSize;
```

Comment:
Do all these variables need to be 16 bit - did not check them in detail.
Saving a few bytes is always good for the smallest of the microprocessors.

Reply:
I had a look at it and the _bufferSize turned out to be redundant.
The _pageSize could be shrunk to 8 bits, as I've only seen 256 byte pages so far and the main code limits it to 128.
But I think the others are as they need to be - even though I get a nagging thought about introducing something that makes it easier for future users to detect if they are trying to use an eeprom that would exceeds the current limits...

Reply:
OK, you gave them a serious second thought, and when needed they can still be updated.